source/extensions/dynamic_modules/abi.h

```diff
@@ -345,6 +378,135 @@ envoy_dynamic_module_on_http_filter_response_trailers(
+    size_t index);
+
+/**
+ * envoy_dynamic_module_callback_http_get_request_trailer_value is exactly the same as the
+ * envoy_dynamic_module_callback_http_get_request_header_value, but for the request trailers.
+ * See the comments on envoy_dynamic_module_http_get_request_header_value for more details.
+ */
+size_t envoy_dynamic_module_callback_http_get_request_trailer_value(
+    envoy_dynamic_module_type_http_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_buffer_module_ptr key, size_t key_length,
+    envoy_dynamic_module_type_buffer_envoy_ptr* result_buffer_ptr, size_t* result_buffer_length_ptr,
+    size_t index);
+
+/**
+ * envoy_dynamic_module_callback_http_get_response_header_value is exactly the same as the
+ * envoy_dynamic_module_callback_http_get_request_header_value, but for the response headers.
+ * See the comments on envoy_dynamic_module_callback_http_get_request_header_value for more details.
+ */
+size_t envoy_dynamic_module_callback_http_get_response_header_value(
+    envoy_dynamic_module_type_http_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_buffer_module_ptr key, size_t key_length,
+    envoy_dynamic_module_type_buffer_envoy_ptr* result_buffer_ptr, size_t* result_buffer_length_ptr,
+    size_t index);
+
+/**
+ * envoy_dynamic_module_callback_http_get_response_trailer_value is exactly the same as the
+ * envoy_dynamic_module_callback_http_get_request_header_value, but for the response trailers.
+ * See the comments on envoy_dynamic_module_callback_http_get_request_header_value for more details.
+ */
+size_t envoy_dynamic_module_callback_http_get_response_trailer_value(
+    envoy_dynamic_module_type_http_filter_envoy_ptr filter_envoy_ptr,
+    envoy_dynamic_module_type_buffer_module_ptr key, size_t key_length,
+    envoy_dynamic_module_type_buffer_envoy_ptr* result_buffer_ptr, size_t* result_buffer_length_ptr,
+    size_t index);
+
+/**
+ * envoy_dynamic_module_callback_http_set_request_header is called by the module to set
+ * the value of the request header with the given key. If the header does not exist, it will be
+ * created. If the header already exists, all existing values will be removed and the new value will
+ * be set.
+ *
+ * @param filter_envoy_ptr is the pointer to the DynamicModuleHttpFilter object of the
+ * corresponding HTTP filter.
+ * @param key is the key of the header.
+ * @param key_length is the length of the key.
+ * @param value is the value of the header.
+ * @param value_length is the length of the value.
+ * @return true if the operation is successful, false otherwise.
+ *
+ * Note that this only sets the header to the underlying Envoy object. Whether or not the header is
+ * actually sent to the upstream depends on the phase of the execution and subsequent
+ * filters. In other words, returning true from this function does not guarantee that the header
+ * will be sent to the upstream.
+ */
+bool envoy_dynamic_module_callback_http_set_request_header(
```

Comment:
note: ditto - this is an abstraction over `Envoy::Http::HeaderMap::set` and the only failure mode is the case where the underlying map is not found, which is covered by this boolean status. when the necessity comes up later, we can switch to defining a dedicated enum status, but for now we should be fine with this (we can break the ABI anyways).